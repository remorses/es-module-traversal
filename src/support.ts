import path from 'path'
import resolve from 'resolve'
import { promises as fsp } from 'fs'
import slash from 'slash'
import { JS_EXTENSIONS, PACKAGE_NAME } from './constants'

export const queryRE = /\?.*$/
export const hashRE = /#.*$/

export const cleanUrl = (url: string) => {
    return url.replace(hashRE, '').replace(queryRE, '')
}

export const debug = require('debug')(PACKAGE_NAME)

export let isRunningWithYarnPnp: boolean
try {
    isRunningWithYarnPnp = Boolean(require('pnpapi'))
} catch {}

export const sleep = (t) => new Promise((res) => setTimeout(res, t))

export async function defaultReadFile(filePath: string): Promise<string> {
    return await (await fsp.readFile(filePath)).toString()
}

export const defaultResolver = (root: string, id: string) => {
    try {
        return (
            resolve.sync(id, {
                basedir: root,
                extensions: [...JS_EXTENSIONS],
                // necessary to work with pnpm
                preserveSymlinks: isRunningWithYarnPnp || false,
            }) || ''
        )
    } catch (e) {
        console.error(`WARN: cannot resolve '${id}' from '${root}'`)
        return ''
    }
}

// this map has same signature as batchedPromiseAll in case i want to refactor and make resolver asynchronous
// export const map = <T, Z>(x: T[], func: (x: T) => Z, _n?: number): Z[] => {
//     return x.map(func)
// }

export function isRelative(x: string) {
    x = cleanUrl(x)
    return x.startsWith('.') || x.startsWith('/')
}

export function isJsModule(x: string) {
    const ext = path.posix.extname(cleanUrl(x))
    return !ext || JS_EXTENSIONS.has(ext)
}

export function flatten<T>(arr: T[][]): T[] {
    return arr.reduce(function (flat, toFlatten) {
        return flat.concat(
            Array.isArray(toFlatten) ? flatten(toFlatten as any) : toFlatten,
        )
    }, [])
}
