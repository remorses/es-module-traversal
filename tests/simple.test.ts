import { init } from 'es-module-lexer'
import path from 'path'
import { URL } from 'url'
import fetch from 'node-fetch'
var serveStatic = require('serve-static')

import os from 'os'
import {
    defaultReadFile,
    defaultResolver,
    isRelative,
    traverseEsModules,
} from '../src'
import { serve } from './support'

it('works', async () => {
    await init
    const currentFile = path.resolve(__dirname, __filename)
    const res = await traverseEsModules({
        entryPoint: currentFile,
    })
})

it('example', async () => {
    const currentFile = path.resolve('tests/example/entry.js')
    const res = await traverseEsModules({
        entryPoint: currentFile,
    })
    expect(res.map((x) => x.importPath)).toMatchSnapshot()
})
it('with server', async () => {
    const PORT = '7000'
    const stop = await serve(PORT)
    const currentFile = path.resolve('tests/example/entry.js')
    const res = await traverseEsModules({
        entryPoint: currentFile,
        resolver: (ctx, importPath) => {
            let importerDirectory = ctx.startsWith('http')
                ? relativePathFromUrl(ctx)
                : ctx
            if (!isRelative(importPath)) {
                return defaultResolver(
                    path.resolve(importerDirectory),
                    importPath,
                )
            }
            importerDirectory = path.relative(process.cwd(), importerDirectory)
            importPath = path.join(importerDirectory, importPath)
            // console.log({ importPath, pathname: importerDirectory })
            return `http://localhost:${PORT}/${importPath}`
        },
        readFile: async (url) => {
            console.log({ url })
            if (!url.startsWith('http')) {
                return defaultReadFile(url)
            }
            // read from the server
            const res = await fetch(url, { headers: {} })
            const content = await res.text()
            return content
        },
    })
    expect(res.map((x) => x.importPath)).toMatchSnapshot()
    await stop()
})

function relativePathFromUrl(ctx) {
    let pathname = new URL(ctx).pathname
    pathname = pathname.startsWith('/') ? pathname.slice(1) : pathname
    return pathname
}
