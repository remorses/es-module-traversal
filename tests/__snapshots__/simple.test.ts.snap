// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example 1`] = `
Array [
  "./firstLevelDynamic.js",
  "./firstLevel.js",
  "./deeper/secondLevel.js",
  "es-module-lexer",
]
`;

exports[`with server 1`] = `
Array [
  "./firstLevelDynamic.js",
  "./firstLevel.js",
  "./deeper/secondLevel.js",
  "es-module-lexer",
]
`;
